_posts/2024/2024-02-26-validate-email-addresses.md

```diff
@@ -0,0 +1,134 @@
+    uniqueness: { case_insensitive: true }
+end
+```
+
+### `URI`
+
+There has been a regular expression built into the `URI` of Ruby‘s standard library since version 2.2. Here's the [commit adding it](https://github.com/ruby/ruby/commit/e63ab5d3ad289767eab49787e4e33390b0ce74e1).
+
+```ruby
+class User < ApplicationRecord
+  validates :email,
+    format: { with: URI::MailTo::EMAIL_REGEXP },
+    presence: true,
+    uniqueness: { case_insensitive: true }
+end
+```
+
+It's been in some of the [Rails Guides examples](https://guides.rubyonrails.org/active_record_validations.html#custom-validators).
+
+### `email_address` gem
+
+The validator from the [`email_address` gem](https://github.com/afair/email_address) has other useful enhancements for working with email.
+
+You can use the built-in validator:
+
+```ruby
+class User < ApplicationRecord
+  validates :email,
+    presence: true,
+    uniqueness: { case_insensitive: true }
+
+  # attribute needs to be called :email
+  validates_with EmailAddress::ActiveRecordValidator
+end
+```
+
+Or directly use the gem‘s functionality in a `validate` block:
+
+```ruby
+class User < ApplicationRecord
+  validates :email,
+    presence: true,
+    uniqueness: { case_insensitive: true }
+
+  validate do |user|
+    EmailAddress::Address.new(user.email).valid?
+  end
+end
+```
+
+Or use a [custom validator](https://guides.rubyonrails.org/active_record_validations.html#custom-validators):
+
+```ruby
+class User < ApplicationRecord
+  validates :email,
+    email: true,
+    presence: true,
+    uniqueness: { case_insensitive: true }
+
+  private
+
+  # You can put this in its own file if you check elsewhere for email validity
+  class EmailValidator < ActiveModel::EachValidator
+    def validate_each(record, attribute, value)
+      unless EmailAddress::Address.new(value).valid?
+        record.errors.add attribute, (options[:message] || "is not an email")
+      end
+    end
+  end
+end
+```
+
+## Why?
+
+Email formatting is a solved problem. _Re_-solving solved problems always feels like madness to me. Why would you not lean on the experience of the entire Internet for something as common as this?
+
+There are various levels of strictness in the regular expressions that underpin the different options I listed above. Most of them are "healthily conventional” in comparison to the [official RFC](https://github.com/afair/email_address?tab=readme-ov-file#avoiding-the-bad-parts-of-rfc-specification) format of emails, which can allow overly unrealistic, but RFC-valid, machine generated email addresses.
+
+If I were to pick, I’d choose the straightforward `URI` option. The underlying regular expression is sufficiently flexible for all regular use.
+
+However, if I was looking to use the [extra functionality](https://github.com/afair/email_address?tab=readme-ov-file#usage) of the `email_address` gem—domain extraction or detection, normalisation or deconstruction—I’d select one of the approaches using that gem.
```

Comment:
This still feels a bit clunky/confusing.

Reply:
```suggestion
If my application needed further email manipulation functionality—detection, normalisation or deconstruction–I'd be tempted to include the `email_address` gem and use it for validation too.
```